### ISSUE TITLE
Some CSV files include a BOM in the three first bytes. Proposal: skip BOMs, or give the option to skip BOMs?

### ISSUE THREAD

**Issue by sroccaserra** [kellyjonbrazil/jc#307]
For some (probably historical) reason, some tools require that UTF-8 encoded CSV files start with a BOM for their three first bytes (Excel :eyes: ?). As a result, many tools produce CSV files that start with the three bytes `0xefbbbf`.

These three chars are not printable, and are not seen in the output, but still present.

This results in the following BOM-plagged file (BOM not shown):

```
a;b
0;1
```
to result in what seems like `[{"a":"0","b":"1"}]`, but in reality the first key is not `a`, which should look like `0x61`, it is `0xefbbbf61`. But it prints just like `a`.

So if you pipe the output of `jc --csv` of this file to the jq query `jq '.[].a'`, it results in `null` (it should result in `"0"`). And it is hard to see why without knowing about BOMs.

Proposal: ignore the first three bytes if they match `0xefbbbf` in csv files (maybe other formats?), or add an option to ignore them if present?

**Comment by kellyjonbrazil** [comment]
Hi - thanks for reporting this! It looks like there may be a clean way to do this in python with 'utf-8-sig' encoding/decoding. I'll investigate...

**Comment by kellyjonbrazil** [comment]
I have a fix in the `dev` branch that seems to fix the issue and doesn't break existing tests:

https://github.com/kellyjonbrazil/jc/commit/888b6bd6d536b003a8a014ef7605fe6e1c918261

With a UTF-8 csv file with BOM (before):
```
$ cat utf-8-bom.csv                         
﻿col1,col2,col3
a,b,c
x,y,z
abc,def,ghi
1,2,3
foo,bar,baz
$ cat utf-8-bom.csv| jc --csv | jq '.[].col1'
null
null
null
null
null
$ cat utf-8-bom.csv| jc --csv-s | jq '.col1' 
null
null
null
null
null
```
After the fix:
```
$ cat utf-8-bom.csv| jc --csv | jq '.[].col1'
"a"
"x"
"abc"
"1"
"foo"
$ cat utf-8-bom.csv| jc --csv-s | jq '.col1' 
"a"
"x"
"abc"
"1"
"foo"
```

**Comment by sroccaserra** [comment]
This is quick, thank you. Also, thank you for jc in the first place, sorry, I was absorbed by the description of the issue.

I am not familiar with the code & the actual perf bottlenecks of jc if any, could the additional encoding / decoding operations cause a peformance drop in big CSVs?

**Comment by kellyjonbrazil** [comment]
You bet - it's my labor of love! 😃 

I should check for the standard `csv` parser. Since it's a lower-level C operation, I'm not anticipating a big difference, but I have some large CSV files I can check against.

I don't anticipate a performance issue with the streaming `csv-s` parser since it only does the encode/decode on the first line. But I'll check that as well.

If there is a performance issue I can find another way to clear the bytes.

**Comment by sroccaserra** [comment]
Hey, that's awesome. Thanks!

**Comment by kellyjonbrazil** [comment]
Not really noticing a significant performance difference parsing a 100k row CSV file:

<img width="805" alt="image" src="https://user-images.githubusercontent.com/24423665/197875093-b98356d6-f145-4f9c-a086-0ece5ccb3b98.png">

Looks like this fix may be ok as-is. If a performance issue is found in the future I can probably figure out a better way.

**Comment by kellyjonbrazil** [comment]
This fix will go in the next release. If you would like to use it in the meantime, you can copy the parser files to your `jc` plugin folder and the original parsers will be overridden:

https://github.com/kellyjonbrazil/jc/blob/dev/jc/parsers/csv.py
https://github.com/kellyjonbrazil/jc/blob/dev/jc/parsers/csv_s.py

Custom local parser plugins may be placed in a `jc/jcparsers` folder in your local "App data directory":

- Linux/unix: `$HOME/.local/share/jc/jcparsers`
- macOS: `$HOME/Library/Application Support/jc/jcparsers`
- Windows: `$LOCALAPPDATA\jc\jc\jcparsers`

**Comment by sroccaserra** [comment]
Thank you, I tested the fix as local parser plugins on my Ubuntu 22.04, Python 3.10.6, it solves my initial issue :tada:

Am I responsible for closing the issue, or should I let you do it?

**Comment by kellyjonbrazil** [comment]
Nice! No worries - I'll close the issue once I have released `jc` v1.22.2 with the fix. Thanks again!

**Comment by kellyjonbrazil** [comment]
Released in `jc` v1.22.2.

**Comment by jbrains** [comment]
I fell into this pit today and I would like you thank you for fixing the problem 3 months before I experienced it.
